// elemental/lib/components/Card.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _constants = require('../constants');

var _constants2 = _interopRequireDefault(_constants);

var React = require('react/addons');

module.exports = React.createClass({
	displayName: 'Card',
	propTypes: {
		children: React.PropTypes.node.isRequired,
		style: React.PropTypes.object
	},
	render: function render() {
		var style = {
			backgroundColor: 'white',
			borderRadius: _constants2['default'].borderRadius.sm,
			boxShadow: '0 2px 3px rgba(0, 0, 0, 0.075), 0 0 0 1px rgba(0,0,0,0.1)',
			marginBottom: _constants2['default'].spacing.md,
			padding: _constants2['default'].spacing.md
		};
		return React.createElement('div', _extends({}, this.props, { style: _extends(style, this.props.style) }));
	}
});
