// elemental/lib/components/Container.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _reactAddons = require('react/addons');

var _reactAddons2 = _interopRequireDefault(_reactAddons);

var _blacklist = require('blacklist');

var _blacklist2 = _interopRequireDefault(_blacklist);

var _constants = require('../constants');

var _constants2 = _interopRequireDefault(_constants);

module.exports = _reactAddons2['default'].createClass({
	displayName: 'Container',
	propTypes: {
		children: _reactAddons2['default'].PropTypes.node.isRequired,
		clearfix: _reactAddons2['default'].PropTypes.bool,
		gutter: _reactAddons2['default'].PropTypes.number,
		maxWidth: _reactAddons2['default'].PropTypes.number,
		style: _reactAddons2['default'].PropTypes.object
	},
	getDefaultProps: function getDefaultProps() {
		return {
			gutter: _constants2['default'].width.gutter,
			maxWidth: _constants2['default'].width.container
		};
	},
	render: function render() {
		var _props = this.props;
		var gutter = _props.gutter;
		var maxWidth = _props.maxWidth;

		var containerStyle = {
			marginLeft: 'auto',
			marginRight: 'auto',
			paddingLeft: gutter,
			paddingRight: gutter,
			maxWidth: maxWidth
		};
		var clearfixStyle = { clear: 'both', display: 'table' };
		var props = (0, _blacklist2['default'])(this.props, 'gutter', 'maxWidth', 'style');

		return this.props.clearfix ? _reactAddons2['default'].createElement(
			'div',
			_extends({ style: _extends(containerStyle, this.props.style) }, props),
			_reactAddons2['default'].createElement('span', { style: clearfixStyle }),
			this.props.children,
			_reactAddons2['default'].createElement('span', { style: clearfixStyle })
		) : _reactAddons2['default'].createElement('div', _extends({ style: _extends(containerStyle, this.props.style) }, props));
	}
});
