// belle/lib/utils/date-helpers.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports.getWeekArrayForMonth = getWeekArrayForMonth;
exports.getLocaleData = getLocaleData;

var _configDatePicker = require('../config/datePicker');

/**
 * The function will take a month and year value and will return an array of weeks for that month.
 * Each element in this array will be in-turn an array of days in the week.
 * @param {number} month: the month for which array of weeks is needed
 * @param {number} year: the year for which array of weeks is needed
 * @param {number} firstDayOfWeek: first day of the week in the locale
 * @returns {Array}: Array of weeks in a month, each week is in turn array of days in that week
 */

function getWeekArrayForMonth(month, year, firstDayOfWeek) {
  var monthDay = new Date(year, month, 1);
  // Todo: simplify this calculation of first date
  var firstDate = 1 + firstDayOfWeek - monthDay.getDay();
  firstDate = firstDate <= 1 ? firstDate : firstDate - 7;
  monthDay.setDate(firstDate);
  var lastDate = new Date(year, month + 1, 0);

  var weekArray = [];
  while (monthDay <= lastDate) {
    var newWeek = [];
    for (var dayCounter = 0; dayCounter < 7; dayCounter++) {
      var weekDate = new Date(monthDay.getFullYear(), monthDay.getMonth(), monthDay.getDate());
      newWeek.push(weekDate);
      monthDay.setDate(monthDay.getDate() + 1);
    }
    weekArray.push(newWeek);
  }
  return weekArray;
}

/**
 * Function will return locale data for locale. If data is not available in config files it will return default data.
 * @param locale - locale for which data is needed.
 * @returns {Object}: Object containing locale data.
 */

function getLocaleData(locale) {
  var localeResult = {};
  var lData = undefined;
  if (locale) {
    lData = _configDatePicker.localeData[locale];
  }
  localeResult.monthNames = lData && lData.monthNames ? lData.monthNames : ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];
  localeResult.dayNamesMin = lData && lData.dayNamesMin ? lData.dayNamesMin : ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa'];
  localeResult.firstDay = lData && lData.firstDay ? lData.firstDay : 0;
  localeResult.weekEnd = lData && lData.weekEnd ? lData.weekEnd : 0;
  localeResult.isRTL = lData && lData.isRTL ? lData.isRTL : false;
  return localeResult;
}

var TODAY = new Date();

exports.TODAY = TODAY;
var CURRENT_DATE = TODAY.getDate();

exports.CURRENT_DATE = CURRENT_DATE;
var CURRENT_MONTH = TODAY.getMonth();

exports.CURRENT_MONTH = CURRENT_MONTH;
var CURRENT_YEAR = TODAY.getFullYear();
exports.CURRENT_YEAR = CURRENT_YEAR;
